ext/standard/flock_compat.h

```diff
@@ -35,18 +35,18 @@
 PHPAPI int php_flock(int fd, int operation);
 
 #ifndef HAVE_FLOCK
-#	define LOCK_SH 1
-#	define LOCK_EX 2
-#	define LOCK_NB 4
-#	define LOCK_UN 8
+#	define LOCK_SH (1 << 0)
+#	define LOCK_EX (1 << 1)
+#	define LOCK_NB (1 << 2)
+#	define LOCK_UN (1 << 3)
 PHPAPI int flock(int fd, int operation);
 #endif
 
 /* Userland LOCK_* constants */
-#define PHP_LOCK_SH 1
-#define PHP_LOCK_EX 2
-#define PHP_LOCK_UN 3
-#define PHP_LOCK_NB 4
+#define PHP_LOCK_SH (1 << 0)
+#define PHP_LOCK_EX (1 << 1)
+#define PHP_LOCK_UN ((1 << 0) | (1 << 1))
```

Comment:
Thank you! It's a bit unfortunate that `PHP_LOCK_UN` looks like an option combining `PHP_LOCK_SH` and `PHP_LOCK_EX` when it's really just a different option. Maybe this could be made more obvious by keeping these options continuous integers.

```php
#define PHP_LOCK_SH 1
#define PHP_LOCK_EX 2
#define PHP_LOCK_UN 3
/* Can be combined with all of the above */
#define PHP_LOCK_NB (1 << 2)
```

It's unfortunate that we have a different binary representation here.

Reply:
unfortunate indeed, as it breaks code like https://3v4l.org/0XT2P 
(which is perfectly valid code in most other languages, like C)

>Maybe this could be made more obvious by keeping these options continuous integers

they shouldn't continuous integers, they should be continuous bits, because they *are* in fact bitflag options, as they're supposed to be combined with LOCK_SH | LOCK_NB and  LOCK_EX | LOCK_NB

(but afaik LOCK_UN is not supposed to be combined with LOCK_NB)

Reply:
> they shouldn't continuous integers, they should be continuous bits, because they are in fact bitflag options, as they're supposed to be combined with LOCK_SH | LOCK_NB and LOCK_EX | LOCK_NB

Yes, I agree, they should be. Unfortunately in PHP they are not. 

> unfortunate indeed, as it breaks code like https://3v4l.org/0XT2P (which is perfectly valid code in most other languages, like C)

Yeah this is not great. The check `($flags & ~LOCK_NB) == LOCK_SH` would be correct but obviously suboptimal. It might be worth addressing this in internals and potentially fix it in PHP 8.2.

Reply:
yeah would be nice to get LOCK_UN fixed, but in the meantime i still think this PR should be merged ^^

Reply:
No? This is clearly not a bitmask, so you shouldn't try to make it look like one. This is a two bit lock type and an NB flag -- writing the two-bit type as if it were a combination of independent bits doesn't make any sense.

Reply:
@nikic btw do you have an opinion on this? https://www.reddit.com/r/programminghorror/comments/uapuo7/php_invented_their_own_nonstandard_value_for_lock/?utm_medium=android_app&utm_source=share